File: facebook-core/src/main/java/com/facebook/appevents/AppEventsManager.java
/*
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 * <p>
 * You are hereby granted a non-exclusive, worldwide, royalty-free license to use,
 * copy, modify, and distribute this software in source code or binary form for use
 * in connection with the web services and APIs provided by Facebook.
 * <p>
 * As with any software that integrates with the Facebook platform, your use of
 * this software is subject to the Facebook Developer Principles and Policies
 * [http://developers.facebook.com/policy/]. This copyright notice shall be
 * included in all copies or substantial portions of the software.
 * <p>
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
 * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
 * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.appevents;

import androidx.annotation.RestrictTo;
import com.facebook.appevents.aam.MetadataIndexer;
import com.facebook.appevents.eventdeactivation.EventDeactivationManager;
import com.facebook.appevents.iap.InAppPurchaseManager;
import com.facebook.appevents.ml.ModelManager;
import com.facebook.appevents.restrictivedatafilter.RestrictiveDataManager;
import com.facebook.internal.FeatureManager;
import com.facebook.internal.FetchedAppSettings;
import com.facebook.internal.FetchedAppSettingsManager;
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions;
import com.facebook.internal.qualityvalidation.Excuse;
import com.facebook.internal.qualityvalidation.ExcusesForDesignViolations;

@ExcusesForDesignViolations(@Excuse(type = "MISSING_UNIT_TEST", reason = "Legacy"))
@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
@AutoHandleExceptions
public class AppEventsManager {
  /**
   * Start AppEvents functionality.
   *
   * <p>Note that the function should be called after FacebookSdk is initialized.
   */
  public static void start() {
    FetchedAppSettingsManager.getAppSettingsAsync(
        new FetchedAppSettingsManager.FetchedAppSettingsCallback() {
          @Override
          public void onSuccess(FetchedAppSettings fetchedAppSettings) {
            FeatureManager.checkFeature(
                FeatureManager.Feature.AAM,
                new FeatureManager.Callback() {
                  @Override
                  public void onCompleted(boolean enabled) {
                    if (enabled) {
                      MetadataIndexer.enable();
                    }
                  }
                });

            FeatureManager.checkFeature(
                FeatureManager.Feature.RestrictiveDataFiltering,
                new FeatureManager.Callback() {
                  @Override
                  public void onCompleted(boolean enabled) {
                    if (enabled) {
                      RestrictiveDataManager.enable();
                    }
                  }
                });

            FeatureManager.checkFeature(
                FeatureManager.Feature.PrivacyProtection,
                new FeatureManager.Callback() {
                  @Override
                  public void onCompleted(boolean enabled) {
                    if (enabled) {
                      ModelManager.enable();
                    }
                  }
                });

            FeatureManager.checkFeature(
                FeatureManager.Feature.EventDeactivation,
                new FeatureManager.Callback() {
                  @Override
                  public void onCompleted(boolean enabled) {
                    if (enabled) {
                      EventDeactivationManager.enable();
                    }
                  }
                });

            FeatureManager.checkFeature(
                FeatureManager.Feature.IapLogging,
                new FeatureManager.Callback() {
                  @Override
                  public void onCompleted(boolean enabled) {
                    if (enabled) {
                      InAppPurchaseManager.enableAutoLogging();
                    }
                  }
                });
          }

          @Override
          public void onError() {}
        });
  }
}

File: facebook-core/src/main/java/com/facebook/appevents/AppEventsManager.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.appevents

import androidx.annotation.RestrictTo
import com.facebook.appevents.aam.MetadataIndexer
import com.facebook.appevents.cloudbridge.AppEventsCAPIManager
import com.facebook.appevents.eventdeactivation.EventDeactivationManager
import com.facebook.appevents.iap.InAppPurchaseManager
import com.facebook.appevents.integrity.BlocklistEventsManager
import com.facebook.appevents.integrity.MACARuleMatchingManager
import com.facebook.appevents.ml.ModelManager
import com.facebook.appevents.integrity.ProtectedModeManager
import com.facebook.appevents.integrity.RedactedEventsManager
import com.facebook.appevents.integrity.SensitiveParamsManager
import com.facebook.appevents.restrictivedatafilter.RestrictiveDataManager
import com.facebook.internal.FeatureManager
import com.facebook.internal.FeatureManager.checkFeature
import com.facebook.internal.FetchedAppSettings
import com.facebook.internal.FetchedAppSettingsManager
import com.facebook.internal.FetchedAppSettingsManager.getAppSettingsAsync
import com.facebook.internal.instrument.crashshield.AutoHandleExceptions

@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
@AutoHandleExceptions
object AppEventsManager {
  /**
   * Start AppEvents functionality.
   *
   * Note that the function should be called after FacebookSdk is initialized.
   */
  @JvmStatic
  fun start() {
    getAppSettingsAsync(
        object : FetchedAppSettingsManager.FetchedAppSettingsCallback {
          override fun onSuccess(fetchedAppSettings: FetchedAppSettings?) {
            checkFeature(FeatureManager.Feature.AAM) { enabled ->
              if (enabled) {
                MetadataIndexer.enable()
              }
            }
            checkFeature(FeatureManager.Feature.RestrictiveDataFiltering) { enabled ->
              if (enabled) {
                RestrictiveDataManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.PrivacyProtection) { enabled ->
              if (enabled) {
                ModelManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.EventDeactivation) { enabled ->
              if (enabled) {
                EventDeactivationManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.IapLogging) { enabled ->
              if (enabled) {
                InAppPurchaseManager.enableAutoLogging()
              }
            }
            checkFeature(FeatureManager.Feature.ProtectedMode) { enabled ->
              if (enabled) {
                ProtectedModeManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.MACARuleMatching) { enabled ->
              if (enabled) {
                MACARuleMatchingManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.BlocklistEvents) { enabled ->
              if (enabled) {
                BlocklistEventsManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.FilterRedactedEvents) { enabled ->
              if (enabled) {
                RedactedEventsManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.FilterSensitiveParams) { enabled ->
              if (enabled) {
                SensitiveParamsManager.enable()
              }
            }
            checkFeature(FeatureManager.Feature.CloudBridge) { enabled ->
              if (enabled) {
                AppEventsCAPIManager.enable()
              }
            }
          }

          override fun onError() = Unit
        })
  }
}
